core/services/relay/evm/read/event.go

```diff
@@ -233,54 +235,121 @@ func (b *EventBinding) BatchCall(_ common.Address, _, _ any) (Call, error) {
+			ContractAddress: address,
+			ContractName:    b.contractName,
+			MethodName:      b.eventName,
+			Params:          params,
+			ReturnVal:       into,
+		}, "", false)
 	}
 
 	topicTypeID := codec.WrapItemType(b.contractName, b.eventName, true)
 
 	onChainTypedVal, err := b.toNativeOnChainType(topicTypeID, params)
 	if err != nil {
-		return fmt.Errorf("failed to convert params to native on chain types: %w", err)
+		callErr := newErrorFromCall(err, Call{
+			ContractAddress: address,
+			ContractName:    b.contractName,
+			MethodName:      b.eventName,
+			Params:          params,
+			ReturnVal:       into,
+		}, strconv.Itoa(int(confs)), false)
+
+		return callErr
 	}
 
 	filterTopics, err := b.extractFilterTopics(topicTypeID, onChainTypedVal)
 	if err != nil {
-		return err
+		return newErrorFromCall(err, Call{
+			ContractAddress: address,
+			ContractName:    b.contractName,
+			MethodName:      b.eventName,
+			Params:          params,
+			ReturnVal:       into,
+		}, strconv.Itoa(int(confs)), false)
 	}
 
 	var log *logpoller.Log
 	if len(filterTopics) != 0 {
 		var hashedTopics []common.Hash
 		hashedTopics, err = b.hashTopics(topicTypeID, filterTopics)
 		if err != nil {
-			return err
+			return newErrorFromCall(err, Call{
+				ContractAddress: address,
+				ContractName:    b.contractName,
+				MethodName:      b.eventName,
+				Params:          params,
+				ReturnVal:       into,
+			}, strconv.Itoa(int(confs)), false)
 		}
 
 		if log, err = b.getLatestLog(ctx, address, confs, hashedTopics); err != nil {
-			return err
+			return newErrorFromCall(err, Call{
+				ContractAddress: address,
+				ContractName:    b.contractName,
+				MethodName:      b.eventName,
+				Params:          params,
+				ReturnVal:       into,
+			}, strconv.Itoa(int(confs)), false)
 		}
 	} else {
 		if log, err = b.lp.LatestLogByEventSigWithConfs(ctx, b.hash, address, confs); err != nil {
-			return wrapInternalErr(err)
+			return newErrorFromCall(wrapInternalErr(err), Call{
+				ContractAddress: address,
+				ContractName:    b.contractName,
+				MethodName:      b.eventName,
+				Params:          params,
+				ReturnVal:       into,
+			}, strconv.Itoa(int(confs)), false)
 		}
 	}
 
-	return b.decodeLog(ctx, log, into)
+	if err := b.decodeLog(ctx, log, into); err != nil {
+		callErr := newErrorFromCall(err, Call{
+			ContractAddress: address,
+			ContractName:    b.contractName,
+			MethodName:      b.eventName,
+			Params:          params,
+			ReturnVal:       into,
+		}, strconv.Itoa(int(confs)), false)
+
+		encoded := hex.EncodeToString(log.Data)
+		callErr.Result = &encoded
+
+		return callErr
+	}
+
+	return nil
 }
 
 func (b *EventBinding) QueryKey(ctx context.Context, address common.Address, filter query.KeyFilter, limitAndSort query.LimitAndSort, sequenceDataType any) ([]commontypes.Sequence, error) {
 	if err := b.validateBound(address); err != nil {
-		return nil, err
+		return nil, newErrorFromCall(err, Call{
```

Comment:
defer wrap error

Reply:
Like so: `fmt.Errorf("%w: %s", err, newErrorFromCall(...))`

Reply:
Or do you mean this?

```golang
defer func() {
    // wrap the error here
}
```

Reply:
second one

Reply:
I ran this in go playground and it doesn't work. I don't think we can wrap the error using a defer statement.

```
package main

import (
	"errors"
	"fmt"
)

var ErrBase = errors.New("base error")

func main() {
	fmt.Printf("%+v", wrapErrorInDefer())
}

func wrapErrorInDefer() error {
	var err error

	defer func() {
		err = fmt.Errorf("%w: detail", ErrBase)
	}()

	return err
}
```

Reply:
It works with named return vars

Reply:
Ahhh. ok. I usually avoid those. I'll try that.